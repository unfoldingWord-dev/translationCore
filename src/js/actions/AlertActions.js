import types from './ActionTypes';

export const closeAlert = (id) => ({
  type: types.CLOSE_ALERT,
  id,
});

/**
 * Opens a new alert.
 * @param {string} id - the alert id
 * @param {*} message - the message to display in the alert
 * @param {func} onConfirm - custom confirm handler
 * @param {func} onCancel - custom cancel handler
 * @param {string} confirmText - the confirm button text
 * @param {string} cancelText - the cancel button text
 * @return {*}
 */
export const openAlert = (id, message, {
  onConfirm = null,
  onCancel = null,
  confirmText = null,
  cancelText = null,
} = {}) => ({
  type: types.OPEN_ALERT,
  id,
  message,
  onConfirm,
  confirmText,
  onCancel,
  cancelText,
});

/**
 * Opens a new alert that can be ignored.
 * @param {string} id - the alert id
 * @param {*} message
 * @param {func} onConfirm - custom confirm handler
 * @param {func} onCancel - custom cancel handler
 * @param {func} onIgnore - custom ignore handler
 * @param {string} confirmText - the confirm button text
 * @param {string} cancelText - the cancel button text
 * @param {string} ignoreText - the ignore checkbox text
 * @return {*}
 */
export const openIgnorableAlert = (id, message, {
  onConfirm = null,
  onCancel = null,
  onIgnore = null,
  confirmText = null,
  cancelText = null,
  ignoreText = null,
} = {}) => ({
  type: types.OPEN_ALERT,
  id,
  message,
  onConfirm,
  confirmText,
  onCancel,
  cancelText,
  onIgnore: typeof onIgnore === 'function' ? onIgnore : () => {},
  ignoreText,
});

/**
 * Ignores a specific alert for the rest of the session
 * @param {string} id - the alert id to ignore
 * @param {boolean} ignore - ignores or enables the alert
 * @return {*}
 */
export const ignoreAlert = (id, ignore = true) => ({
  type: types.IGNORE_ALERT,
  id,
  ignore,
});
