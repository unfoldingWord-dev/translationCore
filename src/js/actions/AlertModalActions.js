/**
 * @module Actions/AlertModal
 */

import consts from './ActionTypes';

/**
 * @description opens the alert dialog with the specified alert message.
 * @param {String} alertMessage - message to be displayed inside the alert dialog.
 * @param {Boolean} loading - true displays spinning icon and no action button in dialog.
 * @return {object} action content.
 */
export function openAlertDialog(alertMessage, loading) {
  return {
    type: consts.OPEN_ALERT_DIALOG,
    alertMessage,
    loading,
  };
}

/**
 * @description opens the option dialog with the specified alert message, and options.
 * @param {String} alertMessage - message to be displayed inside the alert dialog.
 * @param {function} callback - callback function that is called when clicking on button.  First parameter is text on button. Exception is when callback2 is given.
 * @param {String} button1Text - button text to show right button.
 * @param {String} button2Text - button text to show left of right button. (optional - if not present button is not added)
 * @param {String} buttonLinkText - button text to show on left link button. (optional - if not present button is not added)
 * @param {function} callback2 - optional callback function that is called user clicks on button2Text.
 * @return {Object} action content.
 */
export function openOptionDialog(alertMessage, callback, button1Text, button2Text, buttonLinkText = null, callback2 = null) {
  return {
    type: consts.OPEN_OPTION_DIALOG,
    alertMessage,
    callback,
    callback2,
    button1Text,
    button2Text,
    buttonLinkText,
  };
}

/**
 * @description closes the alert dialog.
 * @return {object} action content.
 */
export function closeAlertDialog() {
  return { type: consts.CLOSE_ALERT_DIALOG };
}
