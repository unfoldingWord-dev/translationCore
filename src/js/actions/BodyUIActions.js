import {
  getHomeScreenStep,
  getActiveHomeScreenSteps
} from '../selectors';
import types from './ActionTypes';

/**
 * @description toggles the home view based on param.
 * @param {boolean} boolean - true or false either shows or hides it.
 */
export const toggleHomeView = boolean => ({
  type: types.TOGGLE_HOME_VIEW,
  boolean
});

export const toggleWelcomeSplash = () => ({
  type: types.TOGGLE_WELCOME_SPLASH
});

export const goToNextStep = () => {
  return ((dispatch, getState) => {
    const stepIndex = getHomeScreenStep(getState());
    dispatch(goToStep(stepIndex + 1));
  });
};

export const goToPrevStep = () => {
  return ((dispatch, getState) => {
    const stepIndex = getHomeScreenStep(getState());
    dispatch(goToStep(stepIndex - 1));
  });
};

/**
 * Goes to specified step
 * @param {number} stepNumber - Number of step to go to in home stepper
 */
export const goToStep = stepNumber => {
  return ((dispatch, getState) => {
    if (stepNumber >= 0 && stepNumber <= 3) {
      const activeSteps = getActiveHomeScreenSteps(getState());
      if (!activeSteps[stepNumber]) return;
      dispatch({
        type: types.GO_TO_STEP,
        stepIndex: stepNumber
      });
    } else if (stepNumber < 0) {
      console.error('The min number of steps is 0. (0-3)');
    } else {
      console.error('The max number of steps is 3. (0-3)');
    }
  });
};

export const toggleProjectsFAB = () => ({
  type: types.TOGGLE_PROJECTS_FAB
});

export const openOnlineImportModal = () => ({
  type: types.OPEN_ONLINE_IMPORT_MODAL
});

export const closeOnlineImportModal = () => {
  return {
    type: types.CLOSE_ONLINE_IMPORT_MODAL
  };
};

export const openLicenseModal = () => ({
  type: types.OPEN_LICENSE_MODAL
});

export const closeLicenseModal = () => ({
  type: types.CLOSE_LICENSE_MODAL
});

/**
 * @description show or not show dimmed screen.
 * @param {bool} enable dims the screen if enabled otherwise removes the dim
 */
export const dimScreen = (enable=true) => ({
  type: types.SHOW_DIMMED_SCREEN,
  bool: enable
});
