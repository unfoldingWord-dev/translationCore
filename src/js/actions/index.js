/**
 * @author Ian Hoegen
 * @description: This serves as the new store and events api.
 * @usage:
 * Actions.sendAction('YOUR_EVENT', data) // Emits an event of type YOUR_EVENT, adds data to store
 * Actions.getData('YOUR_EVENT') // Retrieves data from the event emitted
 ***********************************************************************************/

const EventEmitter = require('events').EventEmitter;
const Reducers = require('../reducers');

class Actions extends EventEmitter {
  constructor() {
    super();
    this.store = [];
  }
  sendAction(eventName, actionObj) {
    if(Reducers[eventName]) {
      this.store[eventName] = Reducers[eventName](this.store[eventName], actionObj);
    } else {
      this.store[eventName] = actionObj;
    }
    this.emitChange(eventName);
  }

  subscribe(eventName, callback) {
    this.on(eventName, callback);
  }

  unsubscribe(eventName, callback) {
    this.removeListener(eventName, callback);
  }

  emitChange(eventName, data) {
    this.emit(eventName, data);
  }

  getData (actionType) {
    return this.store[actionType];
  }
}

var actions = new Actions();
module.exports = actions;
