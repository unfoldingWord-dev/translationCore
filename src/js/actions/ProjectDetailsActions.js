import React from 'react';
import consts from './ActionTypes';
import path from 'path-extra';
import fs from 'fs-extra';
import ospath from 'ospath';
import * as bibleHelpers from '../helpers/bibleHelpers';
import * as ProjectDetailsHelpers from '../helpers/ProjectDetailsHelpers';
import * as AlertModalActions from "./AlertModalActions";
import {getTranslate} from "../selectors";
import {cancelProjectValidationStepper} from "./ProjectImportStepperActions";
import * as ProjectOverwriteHelpers from "../helpers/ProjectOverwriteHelpers";
import * as ProjectImportFilesystemActions from "./Import/ProjectImportFilesystemActions";
// constants
const INDEX_FOLDER_PATH = path.join('.apps', 'translationCore', 'index');
const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');

/**
 * @description sets the project save location in the projectDetailReducer.
 * @param {String} pathLocation - project save location and/or directory.
 * @return {object} action object.
 */
export const setSaveLocation = pathLocation => {
  return((dispatch) => {
    dispatch({
      type: consts.SET_SAVE_PATH_LOCATION,
      pathLocation
    });
  });
};

export const resetProjectDetail = () => {
  return {
    type: consts.RESET_PROJECT_DETAIL
  };
};

export function setProjectToolGL(toolName, selectedGL) {
  return (dispatch) => {
    if(typeof toolName !== 'string') {
      return Promise.reject(`Expected "toolName" to be a string but received ${typeof toolName} instead`);
    }
    dispatch({
      type: consts.SET_GL_FOR_TOOL,
      toolName,
      selectedGL
    });
  };
}

export function getProjectProgressForTools(toolName) {
  return (dispatch, getState) => {
    const {
      projectDetailsReducer: {
        projectSaveLocation,
        manifest
      }
    } = getState();
    const bookId = manifest.project.id;
    let progress = 0;
    if(typeof toolName !== 'string') {
      return Promise.reject(`Expected "toolName" to be a string but received ${typeof toolName} instead`);
    }
    const pathToCheckDataFiles = path.join(projectSaveLocation, INDEX_FOLDER_PATH, toolName, bookId);
    if (toolName === 'wordAlignment') {
      const pathToWordAlignmentData = path.join(projectSaveLocation, '.apps', 'translationCore', 'alignmentData', bookId);
      progress = ProjectDetailsHelpers.getWordAlignmentProgress(pathToWordAlignmentData, bookId);
    } else {
      progress = ProjectDetailsHelpers.getToolProgress(pathToCheckDataFiles);
    }

    dispatch({
      type: consts.SET_PROJECT_PROGRESS_FOR_TOOL,
      toolName,
      progress
    });
  };
}

/**
 * @description Sends project manifest to the store
 * @param {object} manifest - manifest file of a project.
 * @return {object} action object.
 */
export function setProjectManifest(manifest) {
  return {
    type: consts.STORE_MANIFEST,
    manifest: manifest
  };
}

/**
 * @description adds a new key name to the manifest object
 * @param {String} propertyName - key string name.
 * ex.
 * manifest {
 *  ...,
 *  [propertyName]: 'value',
 *  ...
 * }
 * @param {*} value - value to be saved in the propertyName
 */
export function addObjectPropertyToManifest(propertyName, value) {
  return {
    type: consts.ADD_MANIFEST_PROPERTY,
    propertyName,
    value
  };
}

export function setProjectBookIdAndBookName() {
  return ((dispatch, getState) => {
    const { bookId } = getState().projectInformationCheckReducer;
    const bookName = bibleHelpers.convertToFullBookName(bookId);
    dispatch({
      type: consts.SAVE_BOOK_ID_AND_BOOK_NAME_IN_MANIFEST,
      bookId,
      bookName
    });
  });
}

export function setProjectResourceId() {
  return ((dispatch, getState) => {
    const { resourceId } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_RESOURCE_ID_IN_MANIFEST,
      resourceId
    });
  });
}

export function setProjectNickname() {
  return ((dispatch, getState) => {
    const { nickname } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_NICKNAME_IN_MANIFEST,
      nickname
    });
  });
}

export function setLanguageDetails() {
  return ((dispatch, getState) => {
    const { languageDirection, languageId, languageName } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_LANGUAGE_DETAILS_IN_MANIFEST,
      languageDirection,
      languageId,
      languageName
    });
  });
}

export function updateContributors() {
  return ((dispatch, getState) => {
    const { contributors } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_TRANSLATORS_LIST_IN_MANIFEST,
      translators: contributors
    });
  });
}

export function updateCheckers() {
  return ((dispatch, getState) => {
    const { checkers } = getState().projectInformationCheckReducer;
    dispatch({
      type: consts.SAVE_CHECKERS_LIST_IN_MANIFEST,
      checkers
    });
  });
}

/**
 * Change project name to match spec and handle overwrite warnings
 * @param {String} projectSaveLocation
 * @param {String} newProjectName
 * @return {Promise} - Returns a promise
 */
export function renameProject(projectSaveLocation, newProjectName) {
  return ((dispatch, getState) => {
    return new Promise(async (resolve) => {
      const projectPath = path.dirname(projectSaveLocation);
      const currentProjectName = path.basename(projectSaveLocation);
      const newProjectPath = path.join(projectPath, newProjectName);
      if (!fs.existsSync(newProjectPath)) {
        ProjectDetailsHelpers.updateProjectFolderName(newProjectName, projectPath, currentProjectName);
        dispatch(setSaveLocation(newProjectPath));
        resolve();
      }
      else { // project name already exists
        const translate = getTranslate(getState());
        const cancelText = translate('buttons.cancel_import_button');
        const continueText = translate('buttons.continue_import_button');
        dispatch(
          AlertModalActions.openOptionDialog(translate('projects.cancel_import_alert'),
            (result) => {
              if (result === cancelText) {
                // if 'cancel import' then close
                // alert and cancel import process.
                dispatch(AlertModalActions.closeAlertDialog());
                dispatch(cancelProjectValidationStepper());
              } else {
                // if 'Continue Import' then just close alert
                dispatch(AlertModalActions.closeAlertDialog());
              }
              resolve();
            },
            continueText,
            cancelText
          )
        );
      }
    });
  });
}

/**
 * handle rename prompting
 * @return {function(*, *): Promise<any>}
 */
export function doRenamePrompting() {
  return (async (dispatch, getState) => {
    const { projectDetailsReducer: {projectSaveLocation} } = getState();
    const hasGitRepo = fs.pathExistsSync(path.join(projectSaveLocation, '.git'));
    if (hasGitRepo) {
      // TODO: implement this in DCS renaming PR
      const newProjectName = path.basename(projectSaveLocation);
      dispatch(AlertModalActions.openOptionDialog("Your local project has been named\n  '" + newProjectName + "'.  \nPardon our mess, but DCS renaming to be implemented in future PR", () => {
        dispatch(AlertModalActions.closeAlertDialog());
      } ));
    } else { // no dcs
      dispatch(ProjectDetailsHelpers.showRenamedDialog());
    }
  });
}

/**
 * if project name needs to be updated to match spec, then project is renamed
 * @param {object} results of the renaming
 * @return {function(*, *): Promise<any>}
 */
export function updateProjectNameIfNecessary(results) {
  return (async (dispatch, getState) => {
    results.repoRenamed = false;
    results.repoExists = false;
    results.newRepoName = null;
    const {
      projectDetailsReducer: {manifest, projectSaveLocation}
    } = getState();
    const { repoNeedsRenaming, newRepoExists, newProjectName } = ProjectDetailsHelpers.shouldProjectNameBeUpdated(manifest, projectSaveLocation);
    results.newRepoName = newProjectName;
    if (repoNeedsRenaming) {
      if (newRepoExists) {
        results.repoExists = true;
      } else {
        results.repoRenamed = true;
        await dispatch(renameProject(projectSaveLocation, newProjectName));
      }
    }
  });
}

/**
 * if project name needs to be updated to match spec, then project is renamed
 * @return {function(*, *): Promise<any>}
 */
export function updateProjectNameIfNecessaryAndDoPrompting() {
  return (async (dispatch) => {
    const renamingResults = {};
    await dispatch(updateProjectNameIfNecessary(renamingResults));
    if (renamingResults.repoRenamed) {
      dispatch(doRenamePrompting());
    }
  });
}

/**
 * handles the prompting for overwrite/merge of project
 * @return {Promise} - Returns a promise
 */
export function handleOverwriteWarning(newProjectPath, projectName) {
  return ((dispatch, getState) => {
    return new Promise(async (resolve) => {
      const translate = getTranslate(getState());
      const confirmText = translate('buttons.overwrite_project');
      const cancelText = translate('buttons.cancel_import_button');
      let overwriteMessage = translate('projects.project_overwrite_has_alignment_message');
      if (! fs.existsSync(path.join(newProjectPath, '.apps'))) {
        overwriteMessage = (
          <div>
            <p>{translate('projects.project_already_exists', {'project_name': projectName})}</p>
            <p>{translate('projects.project_overwrite_no_alignment_message', {over_write: confirmText})}</p>
          </div>
        );
      }
      dispatch(
        AlertModalActions.openOptionDialog(overwriteMessage,
          (result) => {
            if (result === confirmText) {
              dispatch(AlertModalActions.closeAlertDialog());
              const oldProjectPath = path.join(PROJECTS_PATH, projectName);
              ProjectOverwriteHelpers.mergeOldProjectToNewProject(oldProjectPath, newProjectPath);
              fs.removeSync(oldProjectPath); // don't need the oldProjectPath any more now that .apps was merged in
              fs.move(newProjectPath, oldProjectPath); // replace it with new project
              resolve(true);
            } else { // if cancel
              dispatch(AlertModalActions.closeAlertDialog());
              resolve(false);
            }
          },
          cancelText,
          confirmText
        )
      );
    });
  });
}

export function updateProjectTargetLanguageBookFolderName() {
  return ((dispatch, getState) => {
    const {
      projectInformationCheckReducer: { bookId },
      projectDetailsReducer: { projectSaveLocation },
      localImportReducer: { oldSelectedProjectFileName }
    } = getState();
    if (!oldSelectedProjectFileName) {
      console.log("no old selected project File Name");
    } else {
      ProjectDetailsHelpers.updateProjectFolderName(bookId, projectSaveLocation, oldSelectedProjectFileName);
    }
  });
}
