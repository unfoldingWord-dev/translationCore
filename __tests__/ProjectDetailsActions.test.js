/* eslint-env jest */
import types from '../src/js/actions/ActionTypes';
import * as actions from '../src/js/actions/ProjectDetailsActions';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import fs from 'fs-extra';
import path from 'path-extra';
import ospath from "ospath";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');

it('setSaveLocation() creates an action to update contributors', () => {
  const store = mockStore({});
  const expectedActions = [{
    type: types.SET_SAVE_PATH_LOCATION,
    pathLocation: 'some/path'
  }];
  store.dispatch(actions.setSaveLocation('some/path'));
  const receivedActions = store.getActions();
  expect(receivedActions).toEqual(expectedActions);
});

it('resetProjectDetail() creates an action to reset the project details', () => {
  const expectedAction = {
    type: types.RESET_PROJECT_DETAIL
  };
  expect(actions.resetProjectDetail())
    .toEqual(expectedAction);
});

describe('getProjectProgressForTools() should create an action to get the project progress for tools', () => {
  // NOTE: we don't need to test the actual progress checking here.
  // progress checking can be tested on it's own.
  const initialState = {
    projectDetailsReducer: {
      projectSaveLocation: '../',
      manifest: {
        project: {
          id: ''
        }
      }
    }
  };
  fs.__setMockFS({
    [path.join(path.homedir(), 'translationCore/resources/grc/bibles/ugnt/v11/index.json')]: {}
  });

  it('should fail if no toolName is given', () => {
    const store = mockStore(initialState);
    return expect(store.dispatch(actions.getProjectProgressForTools())).rejects.toEqual('Expected "toolName" to be a string but received undefined instead');
  });

  it('should give progress for word alignment', () => {
    const store = mockStore(initialState);
    const expectedActions = [
      {"progress": 0, "toolName": "wordAlignment", "type": "SET_PROJECT_PROGRESS_FOR_TOOL"}
    ];
    store.dispatch(actions.getProjectProgressForTools('wordAlignment'));
    const receivedActions = store.getActions();
    expect(receivedActions).toEqual(expectedActions);
  });

  it('should give progress for a tool', () => {
    const store = mockStore(initialState);
    const expectedActions = [
      {"progress": 0, "toolName": "myTool", "type": "SET_PROJECT_PROGRESS_FOR_TOOL"}
    ];
    store.dispatch(actions.getProjectProgressForTools('myTool'));
    const receivedActions = store.getActions();
    expect(receivedActions).toEqual(expectedActions);
  });
});

describe('setProjectToolGL() should create an action to get the project GL for tools', () => {
  const initialState = {
    projectDetailsReducer: {}
  };

  it('should fail if no toolName is given', () => {
    const store = mockStore(initialState);
    return expect(store.dispatch(actions.setProjectToolGL())).rejects.toEqual('Expected "toolName" to be a string but received undefined instead');
  });

  it('should set GL for word alignment', () => {
    const store = mockStore(initialState);
    const expectedActions = [
      {selectedGL:"hi", toolName:"wordAlignment", type:"SET_GL_FOR_TOOL"}
    ];
    store.dispatch(actions.setProjectToolGL('wordAlignment', 'hi'));
    const receivedActions = store.getActions();
    expect(receivedActions).toEqual(expectedActions);
  });
});

it('setProjectManifest() creates an action to set the project manifest', () => {
  const expectedAction = {
    type: types.STORE_MANIFEST,
    manifest: { hello: 'world' }
  };
  expect(actions.setProjectManifest({ hello: 'world' }))
    .toEqual(expectedAction);
});

it('addObjectPropertyToManifest() creates an action to add an object property to the manifest', () => {
  const expectedAction = {
    type: types.ADD_MANIFEST_PROPERTY,
    propertyName: 'key',
    value: { hello: 'world' }
  };
  expect(actions.addObjectPropertyToManifest('key', { hello: 'world' }))
    .toEqual(expectedAction);
});

it('setProjectBookIdAndBookName() creates an action to set the project book id and name', () => {
  const store = mockStore({
    projectInformationCheckReducer: {
      bookId: 'gen'
    }
  });
  const expectedActions = [{
    type: types.SAVE_BOOK_ID_AND_BOOK_NAME_IN_MANIFEST,
    bookId: 'gen',
    bookName: 'Genesis'
  }];
  store.dispatch(actions.setProjectBookIdAndBookName());
  const receivedActions = store.getActions();
  expect(receivedActions).toEqual(expectedActions);
});

it('setProjectResourceId() creates an action to set the resourceId', () => {
  const store = mockStore({
    projectInformationCheckReducer: {
      resourceId: 'ult',
    }
  });
  const expectedActions = [{
    type: types.SAVE_RESOURCE_ID_IN_MANIFEST,
    resourceId: 'ult'
  }];
  store.dispatch(actions.setProjectResourceId());
  const receivedActions = store.getActions();
  expect(receivedActions).toEqual(expectedActions);
});

it('setProjectNickname() creates an action to set the nickname', () => {
  const store = mockStore({
    projectInformationCheckReducer: {
      nickname: 'Unlocked literal translation',
    }
  });
  const expectedActions = [{
    type: types.SAVE_NICKNAME_IN_MANIFEST,
    nickname: 'Unlocked literal translation'
  }];
  store.dispatch(actions.setProjectNickname());
  const receivedActions = store.getActions();
  expect(receivedActions).toEqual(expectedActions);
});

it('setLanguageDetails() creates an action to set the language details',  () => {
  const store = mockStore({
    projectInformationCheckReducer: {
      languageDirection: 'rtl',
      languageId: 'en',
      languageName: 'English'
    }
  });
  const expectedActions = [{
    type: types.SAVE_LANGUAGE_DETAILS_IN_MANIFEST,
    languageDirection: 'rtl',
    languageId: 'en',
    languageName: 'English'
  }];
  store.dispatch(actions.setLanguageDetails());
  const receivedActions = store.getActions();
  expect(receivedActions).toEqual(expectedActions);
});

it('updateContributors() creates an action to update contributors', () => {
  const store = mockStore({
    projectInformationCheckReducer: {
      contributors: ['jon', 'steve']
    }
  });
  const expectedActions = [{
    type: types.SAVE_TRANSLATORS_LIST_IN_MANIFEST,
    translators: ['jon', 'steve']
  }];
  store.dispatch(actions.updateContributors());
  const receivedActions = store.getActions();
  expect(receivedActions).toEqual(expectedActions);
});

it('updateCheckers() creates an action to update checkers', () => {
  const store = mockStore({
    projectInformationCheckReducer: {
      checkers: ['jon', 'steve']
    }
  });
  const expectedActions = [{
    type: types.SAVE_CHECKERS_LIST_IN_MANIFEST,
    checkers: ['jon', 'steve']
  }];
  store.dispatch(actions.updateCheckers());
  const receivedActions = store.getActions();
  expect(receivedActions).toEqual(expectedActions);
});

describe('ProjectDetailsActions.updateProjectNameIfNecessary()', () => {
  const currentProjectName = "fr_ult_eph_book";
  const currentProjectPath = path.join(PROJECTS_PATH, currentProjectName);
  const mockStoreData = {
    projectDetailsReducer: {
      manifest: {
        target_language: {
          id: 'fr',
          name: 'francais',
          direction: 'ltr'
        },
        project: {
          id: 'eph',
          name: 'Ephesians'
        },
        resource: {
          id: 'ult',
          name: 'unfoldingWord Literal Text'
        }
      },
      projectSaveLocation: currentProjectPath
    },
    localImportReducer: {
      selectedProjectFilename: 'SELECTED_PROJECT_NAME'
    }
  };

  beforeEach(() => {
    // reset mock filesystem data
    fs.__resetMockFS();
    // Set up mock filesystem before each test
    fs.__setMockFS({
      [currentProjectPath]: ''
    });
  });

  afterEach(() => {
    // reset mock filesystem data
    fs.__resetMockFS();
  });

  test('does nothing if project name is valid', async () => {
    // given
    const store = mockStore(mockStoreData);

    // when
    await store.dispatch(actions.updateProjectNameIfNecessary());

    // then
    expect(cleanupPaths(store.getActions())).toMatchSnapshot();
    expect(fs.pathExistsSync(currentProjectPath)).toBeTruthy();
  });

  test('does nothing if projectSaveLocation is not set', async () => {
    // given
    const storeData = JSON.parse(JSON.stringify(mockStoreData));
    delete storeData.projectDetailsReducer.projectSaveLocation;
    const store = mockStore(storeData);

    // when
    await store.dispatch(actions.updateProjectNameIfNecessary());

    // then
    expect(cleanupPaths(store.getActions())).toMatchSnapshot();
    expect(fs.pathExistsSync(currentProjectPath)).toBeTruthy();
  });

  test('renames project if lang_id changed', async () => {
    // given
    const newProjectName = "am_ult_eph_book";
    const expectedProjectPath = path.join(PROJECTS_PATH, newProjectName);
    const storeData = JSON.parse(JSON.stringify(mockStoreData));
    storeData.projectDetailsReducer.manifest.target_language.id = 'am';
    const store = mockStore(storeData);

    // when
    await store.dispatch(actions.updateProjectNameIfNecessary());

    // then
    expect(cleanupPaths(store.getActions())).toMatchSnapshot();
    expect(fs.pathExistsSync(currentProjectPath)).not.toBeTruthy();
    expect(fs.pathExistsSync(expectedProjectPath)).toBeTruthy();
  });

  test('renames project if project id changed', async () => {
    // given
    const newProjectName = "fr_ult_tit_book";
    const expectedProjectPath = path.join(PROJECTS_PATH, newProjectName);
    const storeData = JSON.parse(JSON.stringify(mockStoreData));
    storeData.projectDetailsReducer.manifest.project.id = 'tit';
    const store = mockStore(storeData);

    // when
    await store.dispatch(actions.updateProjectNameIfNecessary());

    // then
    expect(cleanupPaths(store.getActions())).toMatchSnapshot();
    expect(fs.pathExistsSync(currentProjectPath)).not.toBeTruthy();
    expect(fs.pathExistsSync(expectedProjectPath)).toBeTruthy();
  });

  test('renames project if resource.id (resourceId) changed', async () => {
    // given
    const newProjectName = "fr_lib_eph_book";
    const expectedProjectPath = path.join(PROJECTS_PATH, newProjectName);
    const storeData = JSON.parse(JSON.stringify(mockStoreData));
    storeData.projectDetailsReducer.manifest.resource.id = 'lib';
    const store = mockStore(storeData);

    // when
    await store.dispatch(actions.updateProjectNameIfNecessary());

    // then
    expect(cleanupPaths(store.getActions())).toMatchSnapshot();
    expect(fs.pathExistsSync(currentProjectPath)).not.toBeTruthy();
    expect(fs.pathExistsSync(expectedProjectPath)).toBeTruthy();
  });

  test('renames project if new project name is different than spec', async () => {
    // given
    const currentProjectPath = path.join(PROJECTS_PATH, "fr_ULT_eph_book");
    const newProjectName = "fr_ult_eph_book";
    const expectedProjectPath = path.join(PROJECTS_PATH, newProjectName);
    fs.moveSync(expectedProjectPath, currentProjectPath); // move to invalid file
    const storeData = JSON.parse(JSON.stringify(mockStoreData));
    storeData.projectDetailsReducer.projectSaveLocation = currentProjectPath;
    const store = mockStore(storeData);

    // when
    await store.dispatch(actions.updateProjectNameIfNecessary());

    // then
    expect(cleanupPaths(store.getActions())).toMatchSnapshot();
    expect(fs.pathExistsSync(currentProjectPath)).not.toBeTruthy();
    expect(fs.pathExistsSync(expectedProjectPath)).toBeTruthy();
  });

  test('does not rename project if new project name is different than spec and we have duplicate', async () => {
    // given
    const currentProjectPath = path.join(PROJECTS_PATH, "fr_ULT_eph_book");
    const newProjectName = "fr_ult_eph_book";
    const expectedProjectPath = path.join(PROJECTS_PATH, newProjectName);
    fs.moveSync(expectedProjectPath, currentProjectPath); // move to invalid file
    fs.copySync(currentProjectPath, expectedProjectPath); // make duplicate
    const storeData = JSON.parse(JSON.stringify(mockStoreData));
    storeData.projectDetailsReducer.projectSaveLocation = currentProjectPath;
    const store = mockStore(storeData);

    // when
    await store.dispatch(actions.updateProjectNameIfNecessary());

    // then
    expect(cleanupPaths(store.getActions())).toMatchSnapshot();
    expect(fs.pathExistsSync(currentProjectPath)).toBeTruthy();
    expect(fs.pathExistsSync(expectedProjectPath)).toBeTruthy();
  });
});

//
// helpers
//

/**
 * remove user specific paths and just get basename
 * @param {Array} actions
 * @return {*}
 */
function cleanupPaths(actions) {
  if (actions && actions.length) {
    for (let action of actions) {
      if ('pathLocation' in action) {
        action.pathLocation = path.basename(action.pathLocation);
      }
    }
  }
  return actions;
}
