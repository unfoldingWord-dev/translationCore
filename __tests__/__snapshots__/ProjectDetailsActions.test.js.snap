// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDetailsActions.updateProjectNameIfNecessaryAndDoPrompting() does not rename project if new project name is different than spec and we have duplicate 1`] = `Array []`;

exports[`ProjectDetailsActions.updateProjectNameIfNecessaryAndDoPrompting() does nothing if project name is valid 1`] = `Array []`;

exports[`ProjectDetailsActions.updateProjectNameIfNecessaryAndDoPrompting() does nothing if projectSaveLocation is not set 1`] = `Array []`;

exports[`ProjectDetailsActions.updateProjectNameIfNecessaryAndDoPrompting() renames project if lang_id changed 1`] = `
Array [
  Object {
    "pathLocation": "am_ult_eph_book",
    "type": "SET_SAVE_PATH_LOCATION",
  },
  Object {
    "alertMessage": "projects.renamed_project",
    "button1Text": "buttons.ok_button",
    "button2Text": undefined,
    "buttonLinkText": null,
    "callback": [Function],
    "callback2": null,
    "type": "OPEN_OPTION_DIALOG",
  },
]
`;

exports[`ProjectDetailsActions.updateProjectNameIfNecessaryAndDoPrompting() renames project if new project name is different than spec 1`] = `
Array [
  Object {
    "pathLocation": "fr_ult_eph_book",
    "type": "SET_SAVE_PATH_LOCATION",
  },
  Object {
    "alertMessage": "projects.renamed_project",
    "button1Text": "buttons.ok_button",
    "button2Text": undefined,
    "buttonLinkText": null,
    "callback": [Function],
    "callback2": null,
    "type": "OPEN_OPTION_DIALOG",
  },
]
`;

exports[`ProjectDetailsActions.updateProjectNameIfNecessaryAndDoPrompting() renames project if project id changed 1`] = `
Array [
  Object {
    "pathLocation": "fr_ult_tit_book",
    "type": "SET_SAVE_PATH_LOCATION",
  },
  Object {
    "alertMessage": "projects.renamed_project",
    "button1Text": "buttons.ok_button",
    "button2Text": undefined,
    "buttonLinkText": null,
    "callback": [Function],
    "callback2": null,
    "type": "OPEN_OPTION_DIALOG",
  },
]
`;

exports[`ProjectDetailsActions.updateProjectNameIfNecessaryAndDoPrompting() renames project if resource.id (resourceId) changed 1`] = `
Array [
  Object {
    "pathLocation": "fr_lib_eph_book",
    "type": "SET_SAVE_PATH_LOCATION",
  },
  Object {
    "alertMessage": "projects.renamed_project",
    "button1Text": "buttons.ok_button",
    "button2Text": undefined,
    "buttonLinkText": null,
    "callback": [Function],
    "callback2": null,
    "type": "OPEN_OPTION_DIALOG",
  },
]
`;
